src/cargo/util/command_prelude.rs

```diff
@@ -986,6 +989,49 @@ pub fn root_manifest(manifest_path: Option<&Path>, gctx: &GlobalContext) -> Carg
     }
 }
 
+pub fn get_targets_from_rustup() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustup")
+        .arg("target")
+        .arg("list")
+        .arg("--installed")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_from_rustc() -> CargoResult<Vec<String>> {
+    let output = std::process::Command::new("rustc")
+        .arg("--print")
+        .arg("target-list")
+        .output()?;
+
+    if output.status.success() {
+        let stdout = String::from_utf8(output.stdout)?;
+        return Ok(stdout.lines().map(|line| line.to_string()).collect());
+    }
+
+    Ok(vec![])
+}
+
+pub fn get_target_triple() -> Vec<clap_complete::dynamic::CompletionCandidate> {
```

Comment:
nit: since the other functions are implementation details of this one, please list this one first